File: 260-single-number-iii/260-single-number-iii.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        # BruteForce
        
        freq = dict()
        
        for x in nums:
            freq[x] = freq.get(x, 0) + 1
        
        ans = []
        
        for key, value in freq.items():
            if value == 1:
                ans.append(key)
        
        return ans

File: 260-single-number-iii/260-single-number-iii.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        return rahulSir(nums);
        // return solve(nums);        
    }
private:
    vector<int> solve(vector<int>& nums) {
        unordered_map<int, int> freq;
        
        for(int x: nums)
            freq[x]++;
        
        vector<int> ans;
        
        for(auto entry: freq)
            if(entry.second == 1)
                ans.push_back(entry.first);
        return ans;
    }
    
    vector<int> rahulSir(vector<int>& nums) {
        int bitmask = 0;
        
        for(int n: nums)
            bitmask ^= n;
        
        int diff = bitmask & (-1 * (unsigned int)bitmask);
        int y = 0;
        
        for(int n: nums) {
            if((n & diff) != 0)
                y ^= n;
        }
        int x = bitmask ^ y;
        
        return {x, y};
    }
};